allo/backend/hls.py

```diff
@@ -471,6 +531,93 @@ def __call__(self, *args, shell=True):
+                cmd = f"cd {self.project}; vitis_hls -f run.tcl"
+                assert len(args) == 0, "csyn mode does not need to pass in arguments"
+                print(
+                    f"[{time.strftime('%H:%M:%S', time.gmtime())}] Begin synthesizing project ..."
+                )
+                if shell:
+                    process = subprocess.Popen(cmd, shell=True)
+                else:
+                    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
+                process.wait()
+                if process.returncode != 0:
+                    raise RuntimeError("Failed to synthesize the design")
+                # only run block design if configs['run_impl'] or mode contains
+                # 'impl'. Use safe checks instead of catching broad exceptions.
+                do_impl = False
+                if isinstance(self.configs, dict):
+                    do_impl = bool(self.configs.get("run_impl", False))
+                if not do_impl and not (
+                    isinstance(self.mode, str) and "impl" in self.mode
+                ):
+                    # return early if not running impl
+                    return
+
+                # vivado block design
+                bd_script = (
+                    self.custom_bd_tcl if self.custom_bd_tcl else "block_design.tcl"
+                )
+                cmd = f"cd {self.project}; vivado -mode batch -source {bd_script}"
+                print(
+                    f"[{time.strftime('%H:%M:%S', time.gmtime())}] Running Vivado Block Design ..."
+                )
+                if shell:
+                    process = subprocess.Popen(cmd, shell=True)
+                else:
+                    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
+                process.wait()
+                if process.returncode != 0:
+                    raise RuntimeError(
+                        "Failed to create block design/generate bitstream"
+                    )
+
+                # produce host code
+                host_code = codegen_pynq_host(
+                    self.top_func_name,
+                    self.module,
+                    self.project,
+                )
+
+                with open(f"{self.project}/host.py", "w", encoding="utf-8") as outfile:
+                    outfile.write(host_code)
+
+                # group .bit, .hwh, host.py -> deploy folder under the project
+                deploy_dir = os.path.join(self.project, "deploy")
+                cmd = (
+                    f"mkdir -p {deploy_dir}; "
+                    f"cp {self.project}/build_vivado/project_1.runs/impl_1/project_1_bd_wrapper.bit {deploy_dir}/{self.top_func_name}.bit; "
+                    f"cp {self.project}/build_vivado/project_1.gen/sources_1/bd/project_1_bd/hw_handoff/project_1_bd.hwh {deploy_dir}/{self.top_func_name}.hwh; "
+                    f"cp {self.project}/host.py {deploy_dir}/host.py"
+                )
```

Comment:
Move codegen_host and file operations under the `__init__` function but not the `__call__` function.